src/framework/constraints.ml

```diff
@@ -1694,6 +1694,162 @@ struct
+        LH.replace global_visited_calls call ();
+        let new_path_visited_calls = LS.add call path_visited_calls in
+        let fundec_e_typeV: V.t = V.relift (`Right fundec_e) in
+        let gmap = G.callGraph (ctx.global (fundec_e_typeV)) in
+        let callers: CallGraphSet.t = CallGraphMap.find (context_e) gmap in
+        CallGraphSet.iter (fun to_call ->
+            iter_call new_path_visited_calls to_call
+          ) callers;
+      end
+    in
+    let gmap = G.callGraph (ctx.global (v)) in
+    CallGraphMap.iter(fun key value ->
+        let call = (v', key) in
+        iter_call LS.empty call
+      ) gmap (* try all fundec + context pairs that are in the map *)
+
+  let query ctx (type a) (q: a Queries.t): a Queries.result =
+    match q with
+    | WarnGlobal v ->
+      (* check result of loop analysis *)
+      if not (ctx.ask Queries.MustTermAllLoops) then
+        AnalysisState.svcomp_may_not_terminate := true;
+      let v: V.t = Obj.obj v in
+      begin match v with
+        | `Left v' ->
+          S.query (conv ctx) (WarnGlobal (Obj.repr v'))
+        | `Right v' -> cycleDetection ctx v v' (* Note: to make it more efficient, one could only execute the cycle detection in case the loop analysis returns true, because otherwise the program will probably not terminate anyway*)
+      end
+    | InvariantGlobal v ->
+      let v: V.t = Obj.obj v in
+      begin match v with
+        | `Left v ->
+          S.query (conv ctx) (InvariantGlobal (Obj.repr v))
+        | `Right v ->
+          Queries.Result.top q
+      end
+    | _ -> S.query (conv ctx) q
+
+  let branch ctx = S.branch (conv ctx)
+  let assign ctx = S.assign (conv ctx)
+  let vdecl ctx = S.vdecl (conv ctx)
+
+  (* c = context
+     t = set of tuples (fundec * context)
+  *)
+  let side_context sideg f c t =
+    if !AnalysisState.postsolving then
+      sideg (V.contexts f) (G.create_callGraph (CallGraphMap.singleton (c) (t)))
+
+  let enter ctx  = S.enter (conv ctx)
+  let paths_as_set ctx = S.paths_as_set (conv ctx)
+  let body ctx = S.body (conv ctx)
+  let return ctx = S.return (conv ctx)
+  let combine_env ctx r fe f args fc es f_ask =
+    if !AnalysisState.postsolving then
+      let c_r: S.C.t = ctx.context () in (*Caller context*)
+      let nodeF = ctx.node in
+      let fd_r : fundec = Node.find_fundec nodeF in (*Caller fundec*)
+      let c_e: S.C.t = Option.get fc in (*Callee context*)
+      let fd_e : fundec = f in (*Callee fundec*)
+      let tup: (fundec * S.C.t) = (fd_r, c_r) in
+      let t = CallGraphSet.singleton (tup) in
+      side_context ctx.sideg fd_e (c_e) t;
+      S.combine_env (conv ctx) r fe f args fc es f_ask
+    else
+      S.combine_env (conv ctx) r fe f args fc es f_ask
+  let combine_assign ctx = S.combine_assign (conv ctx)
+  let special ctx = S.special (conv ctx)
+  let threadenter ctx = S.threadenter (conv ctx)
+  let threadspawn ctx lv f args fctx = S.threadspawn (conv ctx) lv f args (conv fctx)
```

Comment:
This doesn't handle calls that create threads. What happens if a thread recursively creates itself?

Reply:
Currently, this should be covered by the fact that any multi-threaded program is reported as possibly non-terminating. From what you say, I think that we would only fail in not reporting recursion as a cause for non-termination then.